s3api/handlers_buckets.go

```diff
@@ -13,198 +14,134 @@ import (
-	_, err = s3Service.Service.HeadBucketWithContext(r.Context(), &s3.HeadBucketInput{
-		Bucket: req.Bucket,
-	})
-	if err == nil {
-		msg := fmt.Sprintf("%s: Bucket exists and is owned by you.", s3.ErrCodeBucketAlreadyOwnedByYou)
-		w.WriteHeader(http.StatusConflict)
-		w.Write([]byte(msg))
-		return
-	}
+	// setup rollback function list and defer recovery and execution
+	var rollBackTasks []func() error
+	defer func() {
+		if err := recover(); err != nil {
+			log.Errorf("recovering from panic: %s", err)
+			executeRollBack(&rollBackTasks)
+		}
+	}()
 
-	log.Infof("creating bucket: %s", *req.Bucket)
-	bucketOutput, err := s3Service.Service.CreateBucketWithContext(r.Context(), &req)
+	bucketOutput, err := s3Service.CreateBucket(r.Context(), &req)
 	if err != nil {
-		if aerr, ok := err.(awserr.Error); ok {
-			switch aerr.Code() {
-			case s3.ErrCodeBucketAlreadyExists:
-				msg := fmt.Sprintf("%s: %s", s3.ErrCodeBucketAlreadyExists, aerr.Error())
-				w.WriteHeader(http.StatusConflict)
-				w.Write([]byte(msg))
-			case s3.ErrCodeBucketAlreadyOwnedByYou:
-				msg := fmt.Sprintf("%s: %s", s3.ErrCodeBucketAlreadyOwnedByYou, aerr.Error())
-				w.WriteHeader(http.StatusConflict)
-				w.Write([]byte(msg))
-			default:
-				w.WriteHeader(http.StatusBadRequest)
-				w.Write([]byte(err.Error()))
+		handleError(w, err)
+		msg := fmt.Sprintf("failed to create bucket: %s", err)
+		panic(msg)
+	}
+
+	// append bucket delete to rollback tasks
+	rbfunc := func() error {
+		return func() error {
+			if _, err := s3Service.DeleteEmptyBucket(r.Context(), &s3.DeleteBucketInput{Bucket: req.Bucket}); err != nil {
+				return err
 			}
-		} else {
-			w.WriteHeader(http.StatusBadRequest)
-			w.Write([]byte(err.Error()))
-		}
-		return
+			return nil
+		}()
 	}
+	rollBackTasks = append(rollBackTasks, rbfunc)
 
+	// build the default IAM bucket admin policy (from the config and known inputs)
 	defaultPolicy, err := iamService.DefaultBucketAdminPolicy(req.Bucket)
 	if err != nil {
-		// TODO: delete newly created bucket
-		log.Errorf("failed creating default IAM policy for bucket %s: %s", *req.Bucket, err)
-		w.WriteHeader(http.StatusInternalServerError)
-		return
+		msg := fmt.Sprintf("failed creating default IAM policy for bucket %s: %s", *req.Bucket, err.Error())
+		handleError(w, apierror.New(apierror.ErrInternalError, msg, err))
+		panic(msg)
 	}
 
-	policyName := fmt.Sprintf("%s-BucketAdminPolicy", *req.Bucket)
-	log.Infof("creating IAM policy: %s", policyName)
-
-	policyOutput, err := iamService.Service.CreatePolicyWithContext(r.Context(), &iam.CreatePolicyInput{
+	policyOutput, err := iamService.CreatePolicy(r.Context(), &iam.CreatePolicyInput{
 		Description:    aws.String(fmt.Sprintf("Admin policy for %s bucket", *req.Bucket)),
 		PolicyDocument: aws.String(string(defaultPolicy)),
-		PolicyName:     aws.String(policyName),
+		PolicyName:     aws.String(fmt.Sprintf("%s-BktAdmPlc", *req.Bucket)),
```

Comment:
Looking at these appended labels (e.g. `BktAdmPlc`), maybe it's worth putting them in the config file to make it easier to change.